### ISSUE TITLE
[css-flexbox] overflow clip on SVG elements and flex layout

### ISSUE THREAD

**Issue by khushalsagar** [w3c/csswg-drafts#7714]
This issue is related to the change in #7144 which included a change to add `overflow: clip` to svg elements in UA stylesheet. This causes a compat issue with the following test case:

```html
<style>
  div {
    display: flex;
    width: 100px;
    height: 100px;
  }
  svg {
    flex-grow: 1;
  }
</style>
<div>
  <svg width="150" height="150">
    <rect width="150" height="150" fill="green"></rect>
  </svg>
</div>
```

SVG has a min-width/min-height value of auto via default value for these properties. Before the resolution in #7144, SVG also had `overflow: hidden` applied to it via UA stylesheet.

During flex layout, the minimum width computed for this SVG element with these inputs was 0. Theoretically this aligns with the [spec](https://drafts.csswg.org/css-flexbox/#min-size-auto) based on the text here: "the used value of a [main axis](https://drafts.csswg.org/css-flexbox/#main-axis) [automatic minimum size](https://drafts.csswg.org/css-sizing-3/#automatic-minimum-size) on a flex item that is not a [scroll container](https://drafts.csswg.org/css-overflow-3/#scroll-container) is its [content-based minimum size](https://drafts.csswg.org/css-flexbox/#content-based-minimum-size)".  overflow: hidden causes the element to be a scroll container. But it's specifically not the case for SVG and neither of Chrome/Firefox/Safari make the element scrollable. The relevant spec text is [here](https://svgwg.org/svg2-draft/render.html#OverflowAndClipProperties).

With the change in #7144, the SVG element now gets a value of 'clip' for 'overflow'. This value lines up with how 'hidden' is used on these elements in each browser above. But a side-effect of this is that min-width/min-height:auto now uses the content-based-minimum-size on this element which is incompatible with existing behaviour. The options to fix this are:

1. Add min-width/min-height: 0 to SVG to retain the old behaviour. But this would be breaking if a developer stylesheet changed the value of 'overflow' to 'visible' which earlier would've used min-width/min-height auto.
2. Carve a special-case for svg where 'clip' behaves same as other non-visible values when computing min-width/min-height to retain the current behaviour.

Leaning towards 2) since 'overflow: visible' is likely to be a common use by developers for SVG. SVG has respected overflow:visible to not clip content before #7144.

@bfgeek FYI.

**Comment by bfgeek** [comment]
A third option also is to not change the UA-stylesheet for SVGs (they remain `overflow:hidden` in the UA-stylesheet which is the case today).

**Comment by khushalsagar** [comment]
Based on the resolution in #7144, `hidden` on replaced elements (including SVG) should have a used value of `clip`. See comment [here](https://github.com/w3c/csswg-drafts/issues/7144#issuecomment-1235830910).

Given the above, no change to the UA-stylesheet for SVG should behave the same as adding `overflow: clip` to the stylesheet since they result in the same used value. I think the safest option is to special-case `overflow:clip` on SVGs in flex layout to be backwards compatible with the current behaviour.

**Comment by dholbert** [comment]
Quoting one resolution:
> In the flex spec, for auto min sizing, instead of saying scroll container, we read the computed overflow property value

It wasn't explicitly stated, but I think it's implied that we'll need this change for the [automatic minimum size for grid items](https://drafts.csswg.org/css-grid-1/#min-size-auto) as well.

**Comment by emilio** [comment]
@fantasai @tabatkins grid also uses scroll container, do we want the same "look at the computed overflow value" instead?

**Comment by emilio** [comment]
Ah I guess Daniel's comment is exactly my comment too :)

**Comment by bfgeek** [comment]
@emilio @dholbert - We noticed some new tests came in. It looks like you were asserting that we don't read the computed value of the element in grid like we do in flexbox - is that understanding correct?

(also the tests are kinda asserting unspecified things we believe - like if scrollbars are allowed on buttons etc, but we can fix that in the tests separately).

**Comment by dholbert** [comment]
>  It looks like you were asserting that we don't read the computed value of the element in grid like we do in flexbox - is that understanding correct?

Not quite. I think Gecko does treat grid/flex consistently[i] but we do seem to treat them differently from Chromium/WebKit.

Looking at e.g. http://wpt.live/css/css-grid/stretch-grid-item-text-input-overflow.html (one of the recent test changes), I think the idea of the changes was:
(1) overflow:clip and overflow:visible should behave the same, RE impact on automatic minimum size (this part likely isn't controversial)
(2) inherently-scrollable things like `<input>` shouldn't have their min-size be conditioned on their explicit `overflow` value (this is the part where engines differ I think)

During code-review, I only noticed change (1) and didn't think too much about (2), but it does seem like (2) is in contradiction to the resolution here and is a divergence from other browsers, so we may need to rethink/revert that part.

[i] RE treating grid/flex consistently, here's an example:
```
data:text/html,
<style>f { display: flex; width: 1px;}g { display: grid; width: 1px}</style>
<f><input></f>
<f><input style="overflow:auto"></f>
<g><input></g>
<g><input style="overflow:auto"></g>
```

> (also the tests are kinda asserting unspecified things we believe - like if scrollbars are allowed on buttons etc, but we can fix that in the tests separately).

Mmm right, that seems to be https://wpt.fyi/results/css/css-grid/stretch-grid-item-button-overflow.html (which I think is just a button-flavored version of the `text-input` testcase).  I think that one is similarly expecting that `overflow` has no effect on buttons (in terms of creating scrollbars and also in terms of min-sizing impact), but that's perhaps unjustified on both counts.

**Comment by dholbert** [comment]
(I posted https://bugzilla.mozilla.org/show_bug.cgi?id=1871412#c7 to discuss next-steps on our end.)

**Comment by emilio** [comment]
`<input>` has non-scrollable overflow per spec, see https://html.spec.whatwg.org/#form-controls

**Comment by emilio** [comment]
The button case might be different / unspecified tho

**Comment by emilio** [comment]
> `<input>` has non-scrollable overflow per spec, see https://html.spec.whatwg.org/#form-controls

Note that this is a somewhat recent change tho, and in particular note https://github.com/whatwg/html/pull/10025 which is a proposed tweak to that.

**Comment by dholbert** [comment]
I filed [bug 1873301](https://bugzilla.mozilla.org/show_bug.cgi?id=1873301) to track the behavioral difference for button scrollability, and I filed [bug 1873300](https://bugzilla.mozilla.org/show_bug.cgi?id=1873300) on fixing the WPT to not depend on either of the possible behaviors.
